### PR TITLE
[RNMobile] Update "add block" button style in default editor view

### LINKED ISSUE #4692: Mobile: New Default View: Iterations to Inserter
## Description

It's not always clear to users what primary actions are available when they first open the editor. We'd like to iterate on the default view of the editor to help users discover blocks and, therefore, improve the overall writing/onboarding experience.

The first part of this will involve changes to make the block inserter more visible/clearer. 

## Designs

image-12

_The initial proposal from @iamthomasbishop, internal reference: p9ugOq-2jB-p2#new-default-view_

### PR DESCRIPTION
**Pull request by SiobhyB**
Fixes: https://github.com/wordpress-mobile/gutenberg-mobile/issues/4692

## Related PRs and Installable Builds

* `Gutenberg Mobile:` https://github.com/wordpress-mobile/gutenberg-mobile/pull/4705
* `Android installable build:` https://github.com/wordpress-mobile/WordPress-Android/pull/16196
* `iOS installable build:` https://github.com/wordpress-mobile/WordPress-iOS/pull/18233

## What?

This PR seeks to make the block inserter more visible in the "default view" of the editor. For the purposes of this PR, "default view" has been interpreted as the view of the editor when no blocks (including the editor's title) are selected.

## Why?

User research indicates that it's not always clear to users what primary actions are available when they first open the editor. By iterating on the default view of the editor, it's hoped that it will be easier for users to discover blocks and, therefore, improve the overall writing/onboarding experience.

## How?

With the changes from this PR, whenever a block _isn't_ selected the block inserter button will be styled more prominently and with an `Add Blocks` label. The following is a high-level overview of the code changes involved:

### Logic around "isDefaultView" check

A [new `noContentSelected` check](https://github.com/WordPress/gutenberg/blob/ee541300bae7268c34e359b0a7551fe5f2d0493f/packages/edit-post/src/components/header/header-toolbar/index.native.js#L178) has been implemented to determine whether a block or title is currently selected or not. If nothing is selected within the editor, then `noContentSelected` returns as `true`.

### Styling
 
The `noContentSelected` check is utilised to add different styles to the toolbar and block inserter button in cases when it returns as `true`:

* The `header-toolbar` component (the toolbar at the bottom of the editor) has been updated to [increase in height]([header-toolbar](https://github.com/WordPress/gutenberg/tree/82eeb5f8aeeca855849649479ea488f955d5cb45/packages/edit-post/src/components/header/header-toolbar)[header-toolbar](https://github.com/WordPress/gutenberg/tree/82eeb5f8aeeca855849649479ea488f955d5cb45/packages/edit-post/src/components/header/header-toolbar)) when `noContentSelected` is `true`. To enable this, a new `$mobile-header-toolbar-expanded-height` variable [has been added](https://github.com/WordPress/gutenberg/blob/ee541300bae7268c34e359b0a7551fe5f2d0493f/packages/base-styles/_variables.scss#L114) to the global `_variables.scss` file.
* The `inserter` component (the `+` button within the toolbar) has also been updated to [display a text label](https://github.com/WordPress/gutenberg/blob/ee541300bae7268c34e359b0a7551fe5f2d0493f/packages/block-editor/src/components/inserter/index.native.js#L55) and [apply new styles](https://github.com/WordPress/gutenberg/blob/ee541300bae7268c34e359b0a7551fe5f2d0493f/packages/block-editor/src/components/inserter/index.native.js#L48) when `noContentSelected` is `true`.

## Testing Instructions

To view and test the new button's new styles, follow these steps within the WordPress Android or iOS app:

* Open the editor and tap the button. 
* By default the title will be selected (this will be changed in a separate PR). In order to view the new button, close the keyboard so that no area of the editor is selected (block or title). 
* The `+` inserter will take up a bigger amount of the toolbar than previously, and will display `Add Blocks` text alongside it.
* Selecting a block within the editor will revert the inserter back to its previous style. 

As this is a UI change, it'd be helpful to test the following as part of the above flow:

* Appearance on dark mode vs. light mode.
* Animation when the inserter button changes from the default style and back.
* Display on different device sizes, including the portrait and landscape modes.
* Appearance with different translations, including longer translations and RTL languages. _Note, no translations have been added as of yet, so testing will need to be done by manually editing code with a local setup._
* Verify the text that's read by VoiceOver (iOS) or Talkback (Android) when the new expanded button style is shown is clear (with no regressions).

### General

These before/after screenshots show the inserter button (the `+` icon to the left of the bottom toolbar) as it displayed before the changes in this PR and after. Note, the styling in the "after" screenshot will only be visible when no part of the editor is selected (no block or title selection):

| Before  | After |
| ------------- | ------------- |
|  |  |

### Dark mode

| Dark Mode - Default View | Dark Mode - Block View |
| ------------- | ------------- |
|  |  |

### CODE REVIEW

File: packages/block-editor/src/components/inserter/style.native.scss
Comment (jhnstn) [inline packages/block-editor/src/components/inserter/style.native.scss]: minor nitpick: `dark` is a modifier so if we are sticking with BEM naming the selector should be `inserter-menu__add-block-button--dark` so that the modifier is separated by a `--`

Comment (SiobhyB) [reply packages/block-editor/src/components/inserter/style.native.scss]: Thanks Jason! I've gone ahead to update this in https://github.com/WordPress/gutenberg/pull/39726/commits/b0693a7c688387407a394b4d026f82a8f03a1686.

File: packages/block-editor/src/components/inserter/index.native.js
Comment (SiobhyB) [inline packages/block-editor/src/components/inserter/index.native.js]: Interested in thoughts around whether this comment is justified. It felt like the reasoning for separate "default view" variables might not be clear when I was working on this, but perhaps I'm wrong on that.

**Comment by SiobhyB** [comment]
@jhnstn, @iamthomasbishop, the installable builds for Android ([here](https://github.com/wordpress-mobile/WordPress-Android/pull/16196#issuecomment-1081762187)) and iOS ([here](https://github.com/wordpress-mobile/WordPress-iOS/pull/18233#issuecomment-1081738076)) are up and the code's ready for review now. :) 

In addition to the pending questions at https://github.com/wordpress-mobile/gutenberg-mobile/issues/4692#issuecomment-1077829264, I wanted to note that I anticipate there will probably be design changes needed to the button's padding/spacing/look. Looking forward to hearing any feedback!

File: packages/block-editor/src/components/inserter/index.native.js
Comment (jhnstn) [reply packages/block-editor/src/components/inserter/index.native.js]: I like it. "Default view" is a bit vague , which is fine but the comment helps shore up the meaning.

**Comment by jhnstn** [comment]
I ran through the test steps on the iOS build and ran the branch locally on Android. 
Works as expected 🎉 

Looks like we don't have a translation yet for `add content` ( or at least not one for the language I tried, German) so I tried changing the text directly to something longer. Randomly tried Hungarian and noticed it's quite long according to Google translate : `Tartalom hozzáadása lehetőségre` . That length of text looks fine on a Pixel 6
Screenshot_20220329-170803

I then tried on a smaller device (Nexus One emulator) and noticed the block navigation gets pushed off the screen: 

Screenshot_20220329_172442

This feels like an edge case but we might want to set a fixed width of the add content buttons and have the text wrap so that the block navigation is always visible.

I also just noticed from the screenshots that the vertical text alignment seems off with the Hungarian text. I'm wondering if that has to do with the accent marks ?

**Comment by SiobhyB** [comment]
Thanks for this @jhnstn! I hadn't thought of testing longer translations. 🙇‍♀️ 

> This feels like an edge case but we might want to set a fixed width of the add content buttons and have the text wrap so that the block navigation is always visible.

This might be tricky as the toolbar has a [set height of 44px here](https://github.com/WordPress/gutenberg/blob/30f900ec5e925bdd218a8845db533765834e9e0b/packages/edit-post/src/components/header/header-toolbar/style.native.scss#L3), which would need to be changed in order to have the text wrap onto a new line. I'm also wondering whether this is necessarily a blocker as the undo/redo buttons are still accessible via scrolling. If the user's on a smaller device, I'm imagining they'd already have a need to scroll through the toolbar more than users with bigger devices and it may already be part of their flow. That said, I agree it'd be ideal if all the buttons were available without scrolling and am curious to hear a design perspective.

As an aside, the undo/redo buttons are inactive when first opening a post, which would arguably be when the button is most on display. Perhaps there's a case for hiding them from view when inactive?

> I also just noticed from the screenshots that the vertical text alignment seems off with the Hungarian text. I'm wondering if that has to do with the accent marks?

I think this was the result of me using `padding-top` to align the text instead of a more consistent, accurate way. 🤦‍♀️ I've gone ahead to replace `padding-top` with `align-self` in https://github.com/WordPress/gutenberg/pull/39726/commits/38df61c4dd201b39a860f659f8de74720a5b537f to address that.

File: packages/block-editor/src/components/inserter/index.native.js
Comment (SiobhyB) [reply packages/block-editor/src/components/inserter/index.native.js]: Great, thank for the confidence check! 🙌

**Review by jhnstn — APPROVED** [review]
Everything looks good and I didn't find anything while testing. Not sure if we are awaiting any design feedback but I have nothing more to add. Nice work!

**Comment by SiobhyB** [comment]
Yay, thanks @jhnstn! 🎉 I'll wait for @iamthomasbishop to look this over before merging anything, too.

**Comment by iamthomasbishop** [comment]
@SiobhyB I haven't had a chance to do a thorough review yet, but I noticed one thing at first scan. Our original designs had the height of the "default"/resting state slightly taller to accommodate a slightly larger 'add content' button — I believe it was 56px, whereas the existing toolbar is 44px. Example:

<a href="https://cldup.com/rX6p6VeSLD.png"></a>

**Comment by iamthomasbishop** [comment]
@SiobhyB I went through and fine-tuned the sizing/spacing on the button (and as a result, the toolbar). Here's what we're looking at for specs:

## Button States

### Exanded



### Compact



## Prototype

I put together a quick prototype to show what would look like in practice:

https://user-images.githubusercontent.com/1200432/162085726-eec9ecff-d0d2-4c06-80ca-e3cf7d24693a.mp4

**Comment by SiobhyB** [comment]
Thanks so much for this, @iamthomasbishop!

I've gone ahead to add the specific values for height and padding that you've set out. I believe that the values around the fonts (size and family) are the defaults, but please let me know if I'm mistaken on that.

For the background colour: `$blue-wordpress` is currently in use as the existing "collapsed" icon's background-colour, which is what's being used for the expanded version too. Should I update that to `$blue-50` for both the "collapsed" and "expanded" versions?

**Comment by iamthomasbishop** [comment]
> Should I update that to $blue-50 for both the "collapsed" and "expanded" versions?

That'd be great. It just uses the default accent/blue which should be blue-50 in light mode, blue-30 in dark mode.

**Comment by iamthomasbishop** [comment]
@SiobhyB I just noticed something while testing the latest iOS build, apologies if this is a known issue. If I tap on the compact inserter button (i.e. while the keyboard is shown) I'm seeing the button transform momentarily to the expanded state as the inserter bottomSheet enters. Here's an example:

https://user-images.githubusercontent.com/1200432/162212177-a329da23-2a3d-4cee-902b-b3c96f6dbf2c.MOV

I wonder if this is made worse because the transition on the button states is immediate, but in either way I wouldn't expect the state to change in this specific case.

**Comment by SiobhyB** [comment]
@iamthomasbishop, ah, not a known issue, thank you for spotting that! I'll take a look.

**Comment by hypest** [comment]
👋 @SiobhyB ! I gave https://github.com/WordPress/gutenberg/pull/39726/commits/82eeb5f8aeeca855849649479ea488f955d5cb45 a try and looks like I don't see the text on the Inserter button at all, on the demo app. I tried and reverted https://github.com/WordPress/gutenberg/pull/39726/commits/d2822fe0bf8467a6f3b2dbef8b7cfdc1be80abb6 to see if that could be the culprit and indeed, it looks like that. See the screencast below where I do the revert and then undoing the revert:

https://user-images.githubusercontent.com/1032913/162402460-022848e6-cc85-47a8-988e-e8bb379de8fe.mov

File: packages/block-editor/src/components/inserter/index.native.js
Comment (hypest) [inline packages/block-editor/src/components/inserter/index.native.js]: 👋 @SiobhyB , not sure if you have tried this already for fixing the button flashing issue, but wanted to share some of my findings: a bit of promising success by using the `shouldReplaceBlock` state variable in the Inserter class. Specifically, by adding a `&& ! shouldReplaceBlock` conditional. This way, the text label is not shown when the editor is in the process of replacing the default richtext block.

But as I said, that's half-success, since the toolbar styling is not correct, since that's governed by the `HeaderToolbar` component. Not sure yet how to pass such transitional state to the HeaderToolbar yet.

Hope this all helps!

Comment (SiobhyB) [reply packages/block-editor/src/components/inserter/index.native.js]: @hypest, thank you so much for this! 🙇‍♀️ 

I think you're right that the trickiest part to this is with updating the parent component. With that in mind, I've made a change to set `isDefaultView` as an editor setting, which I envisage as being similar to other editor settings that influence the editor's appearance, such as `fixedToolbar`, `reducedUI`, `focusMode`, etc. 

I've been able to get some code that's _functionally_ working with that approach (which I've committed to this PR) but it still has some issues. The biggest of which is: As my latest code changes involve an import of `@wordpress/editor` inside a `@wordpress/block-editor` file, the automated tests fail with the `Cannot read property 'SETTINGS_DEFAULTS' of undefined, js engine: hermes` error (same issue that ended up blocking this PR: https://github.com/WordPress/gutenberg/pull/35021).

I feel like this is on the right track, though, and it's a case of finding a better way to control the new setting's data. I'll keep looking into this, but want to be transparent that I don't think it's something that will be quick for me to get to the bottom of.

**Comment by SiobhyB** [comment]
For posterity, noting that I've converted this PR back to a draft to reflect the fact it's a work in progress following the discovery of a blocking bug.

File: packages/block-editor/src/components/inserter/index.native.js
Comment (hypest) [reply packages/block-editor/src/components/inserter/index.native.js]: > With that in mind, I've made a change to set isDefaultView as an editor setting, which I envisage as being similar to other editor settings that influence the editor's appearance, such as fixedToolbar, reducedUI, focusMode, etc

Hmm, feels to me that the this showing/hiding of the textual part of the inserter button is not a setting per se. I mean, it's not configuration that we are expecting to hold through the lifecycle of the editor instance. Instead, it's state that changes depending on the user/caret focus. Are there other similar examples already in place perhaps, where we use the settings for state like this?

Perhaps what we need to "copy" is the way the settings are implemented (hooks/events/etc) but not piggyback on the settings component itself.

That said, I wonder if @jhnstn you have thoughts on this?

Comment (SiobhyB) [reply packages/block-editor/src/components/inserter/index.native.js]: > Instead, it's state that changes depending on the user/caret focus. Are there other similar examples already in place perhaps, where we use the settings for state like this?

The only other one that comes to mind is that we use the settings to update `impressions`, which is then used to update parts of the native UI. It was seeing [this line of code](https://github.com/WordPress/gutenberg/blob/95cda71f51ccdb2d764d2c142e746517bdd5a2b6/packages/block-editor/src/components/inserter/index.native.js#L210) in the Inserter component that got me thinking of using the approach I've taken in the latest commits. It feels kind of similar to me in that the `impressions` are more of a state than a setting.

> Perhaps what we need to "copy" is the way the settings are implemented (hooks/events/etc) but not piggyback on the settings component itself.

I wonder if there might be a way to use `context` to change this data. 🤔  Looking forward to hearing any ideas from Jason!

**Comment by fluiddot** [comment]
@SiobhyB I've just pushed a fix in https://github.com/WordPress/gutenberg/pull/39726/commits/bc9b4ea72674bbe43d5070deb0239830897374c9 for the visual glitch outlined in https://github.com/WordPress/gutenberg/pull/39726#issuecomment-1091752804. This has implied reverting your latest changes related to using the global editor settings (revert commits: https://github.com/WordPress/gutenberg/pull/39726/commits/24d929482b74c93eda78de3b8c66698d1c507648 https://github.com/WordPress/gutenberg/pull/39726/commits/adf4173cc02b5c09b713cab86243cddb355593d7). If you agree, we can set the PR back to ready for review and I can proceed with reviewing it, wdyt?

**Comment by SiobhyB** [comment]
@fluiddot, that sounds perfect, thank you 🙇‍♀️

File: packages/block-editor/src/components/inserter/index.native.js
Comment (fluiddot) [reply packages/block-editor/src/components/inserter/index.native.js]: This issue was addressed in https://github.com/WordPress/gutenberg/pull/39726/commits/bc9b4ea72674bbe43d5070deb0239830897374c9.

File: packages/edit-post/src/components/layout/index.native.js
Comment (fluiddot) [inline packages/edit-post/src/components/layout/index.native.js:118]: This value takes the default height of the header toolbar, so it doesn't take into account the case when the toolbar is displayed in the expanded mode. As far as I tested, I didn't identify any visual issue regarding this logic, so looks safe to only use that value.

In the future, it would be great to update this calculation, and instead of accessing the height from the styles, retrieve it from the `onLayout` event.

**Review by fluiddot — APPROVED** [review]
LGTM 🎊 !

_Tested on iPhone 11 (iOS 15.4.1) and Samsung Galaxy S20 FE 5G (Android 11)._

**Comment by fluiddot** [comment]
PR check `End-to-End Tests / Admin - 2` is failing consistently, I checked [the latest commit from `trunk` branch](https://github.com/WordPress/gutenberg/commit/4ed8a414d6314932046de99376ab49c98e129eb9), and looks like the same job failed. Probably the failure is not caused by these changes but we should verify this issue before merging.

**Comment by fluiddot** [comment]
> PR check `End-to-End Tests / Admin - 2` is failing consistently, I checked [the latest commit from `trunk` branch](https://github.com/WordPress/gutenberg/commit/4ed8a414d6314932046de99376ab49c98e129eb9), and looks like the same job failed. Probably the failure is not caused by these changes but we should verify this issue before merging.

I'm going to update the PR with `trunk` as it looks like the PR check that is failing is already fixed per this commit: https://github.com/WordPress/gutenberg/commit/eb61ff9aa493ed6d9dabc4926a516cb65b44689e.

**Pull request merged**
